_sips/sips/2019-08-11-curried-varargs.md

```diff
@@ -0,0 +1,280 @@
+---
+layout: sips
+discourse: true
+title: SIP-NN - Curried varargs
+---
+
+**By: Yang, Bo**
+
+## History
+
+| Date          | Version       |
+|---------------|---------------|
+| Aug 11th 2019 | Initial Draft |
+
+## Introduction
+
+The [repeated parameters](https://scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html#repeated-parameters) syntax is widely used in Scala libraries to create collection initializers, string interpolations, and DSLs. Unfortunately, repeated parameters are type unsafe as it erase all arguments to their common supertype, inefficient as it creates a temporary `Seq` that is difficult to be eliminated by optimizer. In practice, all sophisticated string interpolation libraries, including [string formatting](https://github.com/scala/scala/blob/43e040ff7e4ba92ccf223e77540580b32c1473c0/src/library/scala/StringContext.scala#L94) and [quasiquotes](https://github.com/scala/scala/blob/43e040ff7e4ba92ccf223e77540580b32c1473c0/src/reflect/scala/reflect/api/Quasiquotes.scala#L28) in standard library, [scalameta](https://scalameta.org/docs/trees/quasiquotes.html) and my [fastring](https://github.com/Atry/fastring/blob/67ae4eccdb9b7f58416ed90eae85ddb035b1ffb1/shared/src/main/scala/com/dongxiguo/fastring/Fastring.scala#L242) library, are written in macros in order to avoid runtime overhead of repeated parameters.
+
+We propose **curried varargs** to improve both the type safety and the performance. Given a function call `f(a, b, c)`, when `f` is a subtype of `Curried`, the function call should be rewritten to `f.applyBegin.applyNext(a).applyNext(b).applyNext(c).applyEnd`.
+
+## Motivating Examples
+
+### Examples
+
+Recently I was working on the implementation of [Pre SIP: name based XML literals](https://contributors.scala-lang.org/t/pre-sip-name-based-xml-literals/2175). During implementing that proposal, I found that the proposal is inefficiency due to repeated parameters, and it could be improved dramatically with the help of curried functions.
+
+For example, according to the proposal the XML literal `<div title="my-title">line1<br/>line2</div>` will result the following code:
+
+{% highlight scala %}
```

Comment:
any reason not to do the following?

```
`‌`‌`scala
```

Reply:
According to https://github.com/scala/docs.scala-lang/blob/master/_sips/sip-template.md

Reply:
oh, huh.

Reply:
I don't think that applies anymore